### ISSUE TITLE
ARM service call inside node code

### ISSUE THREAD

**Issue by knelasevero** [epsilonorion/roscopter#11]
Hi,

I can arm with no problems by calling rosservice call /arm as a console command.

But I tried to call it inside the code and it does nothing... both /arm and /disarm.

Can someone show me how to properly call it inside the code? I already followed the Service and clients tutorials of the ROS wiki... It should work...

Maybe I am missing something really stupid.

Anyways, thanks in advance.

Regards,
Lucas

**Comment by tonybaltovski** [comment]
Hi Lucas,

I use a node to arm.  Try something long these lines with the Service tutorials.

```
enum Commands {
CMD_LAUNCH = 1,
CMD_LAND = 2,
CMD_ARM = 3,
CMD_DISARM = 4,
CMD_SET_STABILIZE = 5,
CMD_SET_ALT_HOLD = 6,
CMD_SET_AUTO = 7,
CMD_SET_LOITER = 8,
CMD_SET_LAND = 9,
RETURN_RC_CONTROL = 10
};

ros::ServiceClient apm_cmd_ = nh_.serviceClient<roscopter::APMCommand>("command");

if(arm)
{
      roscopter::APMCommand srv;
      srv.request.command = CMD_ARM;
      ROS_INFO("Trying to arm");
      if (apm_cmd_.call(srv))
            ROS_WARN("Arm request complete");
      else
            ROS_ERROR("Arm request incomplete");
}
```

**Comment by knelasevero** [comment]
Thanks!

I will try that first thing tomorrow at the lab.

Hopefully I will close the issue tomorrow as well!

Regards,
Lucas

**Comment by knelasevero** [comment]
Did not work...

got this error:  "error: ‘APMCommand’ is not a member of ‘roscopter’"

:( Any ideas?

**Comment by tonybaltovski** [comment]
Did you forget `#include <roscopter/APMCommand.h>`?

**Comment by knelasevero** [comment]
Always something stupid.... 
Thanks a lot! It worked.

**Comment by mechAndy** [comment]
Did you have to do anything special to get find APMCommand.h? I have been trying something very similar and ROS cannot find APMCommand.h.

**Comment by tonybaltovski** [comment]
@mechAndy Did you add it as a catkin dependency?

**Comment by mechAndy** [comment]
I have roscopter as a catkin dependencey in my CMake file. However, when I have:
#include <roscopter/APMCommand.h> it cannot find it. 

I can arm it using rosservice call /arm when the roscopter_node, updated/created by @cberzan. However, I cannot figure out how to add the service to my node.

**Comment by tonybaltovski** [comment]
These work for me using my fork.  Give them a try.  If possible, consider [mavros](https://github.com/mavlink/mavros) it is actively being developed.

```
cmake_minimum_required(VERSION 2.8.3)
project(apm)

find_package(catkin REQUIRED COMPONENTS roscpp std_msgs geometry_msgs sensor_msgs tf roscopter)

catkin_package(
  CATKIN_DEPENDS roscpp std_msgs tf roscopter sensor_msgs geometry_msgs
)

include_directories(include
  ${catkin_INCLUDE_DIRS}
)

add_executable(${PROJECT_NAME}_test src/${PROJECT_NAME}_test.cpp)
target_link_libraries(${PROJECT_NAME}_test ${catkin_LIBRARIES})
```

```

#include <ros/ros.h>
#include <sensor_msgs/Joy.h>
#include <roscopter/APMCommand.h>

enum Commands
{
  CMD_LAUNCH = 1,
  CMD_LAND = 2,
  CMD_ARM = 3,
  CMD_DISARM = 4,
  CMD_SET_STABILIZE = 5,
  CMD_SET_ALT_HOLD = 6,
  CMD_SET_AUTO = 7,
  CMD_SET_LOITER = 8,
  CMD_SET_LAND = 9,
  RETURN_RC_CONTROL = 10
};

enum Modes
{
  CONTROL_DISABLED = 0,
  TELEOP = 1,
  HOVER_CONTROL = 2
};



class APMTest
{
private:
  // Node handle
  ros::NodeHandle nh_;
  // Services
  ros::ServiceClient apm_cmd_;
  // Subscribers
  ros::Subscriber joy_sub_;
  // Members
  int arm_button_;
  int disarm_button_;

public:
  APMTest() :
    // Members default values
    arm_button_(2),
    disarm_button_(3)
  {
    // ROS params
    nh_.param("arm_button", arm_button_, arm_button_);
    nh_.param("disarm_button", disarm_button_, disarm_button_);
    // Init everything
    init_services();
    init_subscribers();
  }

  void init_services()
  {
    // Init Srvs
    apm_cmd_ = nh_.serviceClient<roscopter::APMCommand>("command");
  }

  void init_subscribers()
  {
    // Init Subs
    joy_sub_ =
                          nh_.subscribe("joy", 1, &APMTest::joy_callback, this,
                          ros::TransportHints().unreliable().reliable().tcpNoDelay());
  }


  void joy_callback(const sensor_msgs::Joy::ConstPtr& joy_msg)
  {
    // Motor arming/disarming
    if (joy_msg->buttons[arm_button_] == 1)
    {
      roscopter::APMCommand srv;
      srv.request.command = CMD_ARM;
      ROS_INFO("Trying to arm");
      if (apm_cmd_.call(srv))
      {
        ROS_WARN("Arm request complete");
      }
      else
      {
        ROS_ERROR("Arm request incomplete");
      }
    }
    else if (joy_msg->buttons[disarm_button_] == 1)
    {
      roscopter::APMCommand srv;
      srv.request.command = CMD_DISARM;
      ROS_INFO("Trying to dis-arm");
      if (apm_cmd_.call(srv))
      {
        ROS_WARN("Dis-arm request complete");
      }
      else
      {
        ROS_ERROR("Dis-arm request incomplete");
      }
    }
  }
};


int main(int argc, char** argv)
{
  ros::init(argc, argv, "apm_test");
  APMTest apm_test;
  ros::spin();
  return 0;
}
```

```
<?xml version="1.0"?>
<package>
  <name>apm</name>
  <version>0.0.2</version>
  <description> apm test </description>
  <maintainer email="tony.baltovski@gmail.com"> Tony Baltovski</maintainer>
  <license>BSD</license>
  <author email="tony.baltovski@gmail.com"> Tony Baltovski</author>

  <buildtool_depend>catkin</buildtool_depend>
  <build_depend>roscpp</build_depend>
  <build_depend>std_msgs</build_depend>
  <build_depend>roscopter</build_depend>
  <run_depend>roscpp</run_depend>
  <run_depend>std_msgs</run_depend>
  <run_depend>joy</run_depend>
  <run_depend>roscopter</run_depend>
</package>
```

**Comment by mechAndy** [comment]
@tonybaltovski - Thanks for the help. I was able to get it up and running. We were thinking about going with the APM 2.6 over the pixhawk. Do you have any experience with the APM 2.6? We are not getting anything back from the heart beat.

**Comment by tonybaltovski** [comment]
I am currently using the APM 2.6.  Can you get it connected using a ground station?

**Comment by mechAndy** [comment]
Yes. It will connect to the ground station but when we directly connect from the odroid to the APM 2.6 we never get a response at the "waiting for heartbeat" stage.

**Comment by tonybaltovski** [comment]
@mechAndy I don't recall having that issue. I tried to compile roscopter but pymavlink isn't released with the mavlink deb (mavlink/mavros#245).  Have you tried connecting to a desktop with roscopter?  Also, try mavros if that doesn't work.
